File: crates/ozzy-core/src/project.rs
//! Project management - ozzy.toml and .ozzy/ directory.
//!
//! A project directory contains:
//! - ozzy.toml: Project metadata and current HEAD
//! - .ozzy/: Internal directory with commits, refs, and objects
//! - data/: Working directory for staged data
//! - transforms/: Working directory for staged transforms

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::{BTreeMap, HashSet};
use std::fs;
use std::path::{Component, Path, PathBuf};

use crate::error::{Error, Result};

/// Validate that a name is safe for use in paths (no path traversal).
/// Names should be simple identifiers without slashes, dots at the start, or other special chars.
pub fn validate_safe_name(name: &str) -> Result<()> {
    if name.is_empty() {
        return Err(Error::InvalidPath("Name cannot be empty".to_string()));
    }

    // Block leading dots (hidden files / parent traversal)
    if name.starts_with('.') {
        return Err(Error::InvalidPath(format!(
            "Name cannot start with a dot: {}",
            name
        )));
    }

    // Allow alphanumeric, underscore, hyphen, and dot (for semver tags like v1.0.0)
    if !name
        .chars()
        .all(|c| c.is_ascii_alphanumeric() || c == '_' || c == '-' || c == '.')
    {
        return Err(Error::InvalidPath(format!(
            "Name must contain only [a-zA-Z0-9._-]: {}",
            name
        )));
    }

    Ok(())
}

/// Validate that a path stays within a base directory after canonicalization.
/// This prevents symlink attacks and path traversal via .. components.
pub fn validate_path_within(path: &Path, base_dir: &Path) -> Result<()> {
    // Canonicalize both paths to resolve symlinks and .. components
    let canonical_base = base_dir.canonicalize().map_err(|e| {
        Error::InvalidPath(format!(
            "Cannot canonicalize base directory {}: {}",
            base_dir.display(),
            e
        ))
    })?;

    let canonical_path = path.canonicalize().map_err(|e| {
        Error::InvalidPath(format!(
            "Cannot canonicalize path {}: {}",
            path.display(),
            e
        ))
    })?;

    // Check that the canonical path starts with the canonical base
    if !canonical_path.starts_with(&canonical_base) {
        return Err(Error::InvalidPath(format!(
            "Path {} escapes base directory {}",
            path.display(),
            base_dir.display()
        )));
    }

    Ok(())
}

/// Project configuration stored in ozzy.toml
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectConfig {
    pub project: ProjectMetadata,

    #[serde(default)]
    pub refs: RefsConfig,

    /// Named remote registries.
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub remotes: BTreeMap<String, String>,

    #[serde(default)]
    pub workspace: WorkspaceConfig,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectMetadata {
    pub name: String,
    pub owner: String,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,

    #[serde(default = "default_version")]
    pub version: String,
}

fn default_version() -> String {
    "0.1.0".to_string()
}

fn ensure_root_gitignore(root: &Path) -> Result<()> {
    let gitignore_path = root.join(".gitignore");
    let required_entries = [".ozzy/", "data/*.parquet"];

    let mut content = if gitignore_path.exists() {
        fs::read_to_string(&gitignore_path)?
    } else {
        String::new()
    };

    let existing: HashSet<String> = content
        .lines()
        .map(str::trim)
        .filter(|line| !line.is_empty())
        .map(|line| line.to_string())
        .collect();

    let missing: Vec<&str> = required_entries
        .iter()
        .copied()
        .filter(|entry| !existing.contains(*entry))
        .collect();

    if missing.is_empty() {
        return Ok(());
    }

    if !content.is_empty() && !content.ends_with('\n') {
        content.push('\n');
    }
    if content.is_empty() {
        content.push_str("# OzzyDB generated\n");
    }
    for entry in missing {
        content.push_str(entry);
        content.push('\n');
    }

    fs::write(gitignore_path, content)?;
    Ok(())
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct RefsConfig {
    /// Current HEAD reference (e.g., "refs/heads/main")
    #[serde(default = "default_head")]
    pub head: String,

    /// Remote registry URL (optional)
    #[serde(skip_serializing_if = "Option::is_none")]
    pub remote: Option<String>,
}

fn default_head() -> String {
    "refs/heads/main".to_string()
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct WorkspaceConfig {
    /// Staged data files
    #[serde(default)]
    pub staged_data: Vec<String>,

    /// Staged transform files
    #[serde(default)]
    pub staged_transforms: Vec<String>,
}

/// Data source metadata
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct DataSource {
    pub name: String,
    pub hash: String,
    pub schema_hash: String,
    pub path: String, // Relative path in project
    pub row_count: Option<u64>,
    pub byte_size: Option<u64>,
}

/// Transform metadata
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Transform {
    pub name: String,
    pub hash: String,          // Composite: blake3(source_hash + function_name + lockfile + runtime + params_schema)
    pub runtime: String,       // e.g., "python-3.11"
    pub source_path: String,   // Relative path in project
    pub function_name: String, // Function to call
    pub lockfile_hash: String,
    pub params_schema: serde_json::Value,

    /// Content hash of the canonical source file: blake3(canonicalize_source(text)).
    /// Used for content-addressed storage lookups and integrity verification.
    /// Distinct from `hash` which is the composite identity hash for caching.
    #[serde(default)]
    pub source_hash: String,

    #[serde(default = "default_reproducible")]
    pub reproducible: bool,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub input_schema: Option<serde_json::Value>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub output_schema: Option<serde_json::Value>,
}

fn default_reproducible() -> bool {
    true
}

/// Pipeline node in an endpoint's DAG
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PipelineNode {
    pub node_name: String,
    pub transform_name: String,
    pub params: serde_json::Value,
}

/// Pipeline edge representing data flow
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct PipelineEdge {
    pub target_node: String,
    pub input_name: String,
    pub source_type: SourceType,
    pub source_ref: String,

    // For external dependencies
    #[serde(skip_serializing_if = "Option::is_none")]
    pub external_owner: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub external_project: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub external_endpoint: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub external_commit_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum SourceType {
    DataSource,
    Node,
    External,
}

/// Endpoint definition
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct Endpoint {
    pub name: String,
    pub nodes: Vec<PipelineNode>,
    pub edges: Vec<PipelineEdge>,

    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
}

/// Commit object
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Commit {
    pub hash: String,
    pub parent_hashes: Vec<String>,
    pub author: String,
    pub message: String,
    pub timestamp: DateTime<Utc>,
    /// Data sources (BTreeMap for deterministic serialization order)
    pub data_sources: BTreeMap<String, DataSource>,
    /// Transforms (BTreeMap for deterministic serialization order)
    pub transforms: BTreeMap<String, Transform>,
    /// Endpoints (BTreeMap for deterministic serialization order)
    pub endpoints: BTreeMap<String, Endpoint>,
}

/// Project handle for interacting with an OzzyDB project
pub struct Project {
    /// Root directory of the project
    pub root: PathBuf,

    /// Project configuration
    pub config: ProjectConfig,
}

impl Project {
    /// Find and open a project in the current directory or parents.
    pub fn find_current() -> Result<Self> {
        let current = std::env::current_dir()?;
        Self::find_in(&current)
    }

    /// Find a project starting from the given directory.
    pub fn find_in(start: &Path) -> Result<Self> {
        let mut dir = start.to_path_buf();
        loop {
            let config_path = dir.join("ozzy.toml");
            if config_path.exists() {
                return Self::open(&dir);
            }
            if !dir.pop() {
                return Err(Error::NotInProject);
            }
        }
    }

    /// Open an existing project at the given path.
    pub fn open(root: &Path) -> Result<Self> {
        let config_path = root.join("ozzy.toml");
        if !config_path.exists() {
            return Err(Error::NotInProject);
        }

        let config_content = fs::read_to_string(&config_path)?;
        let config: ProjectConfig = toml::from_str(&config_content)?;

        Ok(Self {
            root: root.to_path_buf(),
            config,
        })
    }

    /// Initialize a new project.
    pub fn init(root: &Path, name: &str, owner: &str) -> Result<Self> {
        let config_path = root.join("ozzy.toml");
        if config_path.exists() {
            return Err(Error::ProjectAlreadyExists);
        }

        // Create directory structure
        fs::create_dir_all(root)?;
        fs::create_dir_all(root.join(".ozzy/commits"))?;
        fs::create_dir_all(root.join(".ozzy/refs/heads"))?;
        fs::create_dir_all(root.join(".ozzy/refs/tags"))?;
        fs::create_dir_all(root.join(".ozzy/objects/data"))?;
        fs::create_dir_all(root.join(".ozzy/objects/transforms"))?;
        fs::create_dir_all(root.join("data"))?;
        fs::create_dir_all(root.join("transforms"))?;

        // Create initial config
        let config = ProjectConfig {
            project: ProjectMetadata {
                name: name.to_string(),
                owner: owner.to_string(),
                description: None,
                version: default_version(),
            },
            refs: RefsConfig {
                head: default_head(),
                remote: None,
            },
            remotes: BTreeMap::new(),
            workspace: WorkspaceConfig::default(),
        };

        let config_content = toml::to_string_pretty(&config)?;
        fs::write(&config_path, config_content)?;

        // Ensure git ignores local internals and raw parquet inputs.
        ensure_root_gitignore(root)?;

        Ok(Self {
            root: root.to_path_buf(),
            config,
        })
    }

    /// Save the project config to disk.
    pub fn save_config(&self) -> Result<()> {
        let config_path = self.root.join("ozzy.toml");
        let config_content = toml::to_string_pretty(&self.config)?;
        fs::write(config_path, config_content)?;
        Ok(())
    }

    // ─────────────────────────────────────────────────────────────────────
    // Path helpers
    // ─────────────────────────────────────────────────────────────────────

    pub fn ozzy_dir(&self) -> PathBuf {
        self.root.join(".ozzy")
    }

    pub fn commits_dir(&self) -> PathBuf {
        self.ozzy_dir().join("commits")
    }

    pub fn refs_dir(&self) -> PathBuf {
        self.ozzy_dir().join("refs")
    }

    pub fn objects_dir(&self) -> PathBuf {
        self.ozzy_dir().join("objects")
    }

    pub fn data_dir(&self) -> PathBuf {
        self.root.join("data")
    }

    pub fn transforms_dir(&self) -> PathBuf {
        self.root.join("transforms")
    }

    fn normalize_ref_relative_path(ref_name: &str) -> Result<PathBuf> {
        let relative = ref_name.strip_prefix("refs/").unwrap_or(ref_name);
        if relative.is_empty() {
            return Err(Error::InvalidPath("Ref name cannot be empty".to_string()));
        }

        let relative_path = Path::new(relative);
        if relative_path.is_absolute() {
            return Err(Error::InvalidPath(format!(
                "Ref path must be relative: {}",
                ref_name
            )));
        }

        let mut sanitized = PathBuf::new();
        for component in relative_path.components() {
            match component {
                Component::Normal(part) => {
                    let part_str = part.to_str().ok_or_else(|| {
                        Error::InvalidPath(format!(
                            "Ref contains non-UTF-8 path component: {}",
                            ref_name
                        ))
                    })?;
                    validate_safe_name(part_str)?;
                    sanitized.push(part);
                }
                _ => {
                    return Err(Error::InvalidPath(format!(
                        "Invalid ref path '{}': path traversal or absolute components are not allowed",
                        ref_name
                    )));
                }
            }
        }

        if sanitized.as_os_str().is_empty() {
            return Err(Error::InvalidPath("Ref name cannot be empty".to_string()));
        }

        Ok(sanitized)
    }

    fn assert_ref_path_within_refs_dir(&self, ref_path: &Path) -> Result<()> {
        let refs_dir = self.refs_dir();
        let canonical_refs = refs_dir.canonicalize().map_err(|e| {
            Error::InvalidPath(format!(
                "Cannot canonicalize refs directory {}: {}",
                refs_dir.display(),
                e
            ))
        })?;

        if ref_path.exists() {
            let canonical_ref_path = ref_path.canonicalize().map_err(|e| {
                Error::InvalidPath(format!(
                    "Cannot canonicalize ref path {}: {}",
                    ref_path.display(),
                    e
                ))
            })?;
            if !canonical_ref_path.starts_with(&canonical_refs) {
                return Err(Error::InvalidPath(format!(
                    "Ref path {} escapes refs directory {}",
                    ref_path.display(),
                    refs_dir.display()
                )));
            }
            return Ok(());
        }

        let mut current = ref_path.parent().ok_or_else(|| {
            Error::InvalidPath(format!(
                "Ref path has no parent directory: {}",
                ref_path.display()
            ))
        })?;

        loop {
            if current.exists() {
                let canonical_current = current.canonicalize().map_err(|e| {
                    Error::InvalidPath(format!(
                        "Cannot canonicalize ancestor path {}: {}",
                        current.display(),
                        e
                    ))
                })?;
                if !canonical_current.starts_with(&canonical_refs) {
                    return Err(Error::InvalidPath(format!(
                        "Ref path {} escapes refs directory {}",
                        ref_path.display(),
                        refs_dir.display()
                    )));
                }
                return Ok(());
            }

            current = current.parent().ok_or_else(|| {
                Error::InvalidPath(format!(
                    "Ref path {} does not have an ancestor within refs directory",
                    ref_path.display()
                ))
            })?;
        }
    }

    fn checked_ref_path(&self, ref_name: &str) -> Result<PathBuf> {
        let relative = Self::normalize_ref_relative_path(ref_name)?;
        let ref_path = self.refs_dir().join(relative);
        self.assert_ref_path_within_refs_dir(&ref_path)?;
        Ok(ref_path)
    }

    // ─────────────────────────────────────────────────────────────────────
    // Ref operations
    // ─────────────────────────────────────────────────────────────────────

    /// Get the current HEAD commit hash.
    pub fn head_commit(&self) -> Result<Option<String>> {
        self.resolve_ref(&self.config.refs.head)
    }

    /// Resolve a ref to a commit hash.
    ///
    /// Supports multiple formats:
    /// - `refs/heads/main` or `refs/tags/v1.0` — direct ref path lookup
    /// - `@latest` — resolves to the default branch (refs/heads/main)
    /// - `@v1.0.0` — resolves to a tag (refs/tags/v1.0.0)
    /// - 64-char hex string — treated as a raw commit hash
    pub fn resolve_ref(&self, ref_name: &str) -> Result<Option<String>> {
        // Handle @latest → default branch
        if ref_name == "@latest" {
            let default_branch = &self.config.refs.head;
            let ref_path = self.checked_ref_path(default_branch)?;
            if ref_path.exists() {
                let hash = fs::read_to_string(&ref_path)?.trim().to_string();
                return Ok(Some(hash));
            }
            return Ok(None);
        }

        // Handle @tag_name → refs/tags/tag_name
        if let Some(tag_name) = ref_name.strip_prefix('@') {
            let tag_path = self.checked_ref_path(&format!("refs/tags/{}", tag_name))?;
            if tag_path.exists() {
                let hash = fs::read_to_string(&tag_path)?.trim().to_string();
                return Ok(Some(hash));
            }
            return Ok(None);
        }

        // Handle raw commit hash (64-char hex string)
        if ref_name.len() == 64 && ref_name.chars().all(|c| c.is_ascii_hexdigit()) {
            let commit_path = self.commits_dir().join(format!("{}.json", ref_name));
            if commit_path.exists() {
                return Ok(Some(ref_name.to_string()));
            }
            return Ok(None);
        }

        // Standard ref path lookup
        let ref_path = self.checked_ref_path(ref_name)?;
        if ref_path.exists() {
            let hash = fs::read_to_string(&ref_path)?.trim().to_string();
            Ok(Some(hash))
        } else {
            Ok(None)
        }
    }

    /// Update a ref to point to a commit.
    pub fn update_ref(&self, ref_name: &str, commit_hash: &str) -> Result<()> {
        let refs_dir = self.refs_dir();
        fs::create_dir_all(&refs_dir)?;

        let ref_path = self.checked_ref_path(ref_name)?;
        if let Some(parent) = ref_path.parent() {
            fs::create_dir_all(parent)?;
        }
        self.assert_ref_path_within_refs_dir(&ref_path)?;
        fs::write(ref_path, format!("{}\n", commit_hash))?;
        Ok(())
    }

    // ─────────────────────────────────────────────────────────────────────
    // Commit operations
    // ─────────────────────────────────────────────────────────────────────

    /// Load a commit by hash, verifying its integrity.
    pub fn load_commit(&self, hash: &str) -> Result<Commit> {
        let commit_path = self.commits_dir().join(format!("{}.json", hash));
        if !commit_path.exists() {
            return Err(Error::CommitNotFound(hash.to_string()));
        }
        let content = fs::read_to_string(&commit_path)?;
        let commit: Commit = serde_json::from_str(&content)?;

        let expected = crate::commit::compute_commit_hash(&commit);
        if expected != commit.hash {
            return Err(Error::InvalidPath(format!(
                "Commit hash mismatch: file says {}, computed {}. Commit may be corrupted.",
                commit.hash, expected
            )));
        }

        Ok(commit)
    }

    /// Save a commit to disk.
    pub fn save_commit(&self, commit: &Commit) -> Result<()> {
        let commit_path = self.commits_dir().join(format!("{}.json", commit.hash));
        let content = serde_json::to_string_pretty(commit)?;
        fs::write(commit_path, content)?;
        Ok(())
    }

    /// Get the latest commit (from HEAD).
    pub fn latest_commit(&self) -> Result<Option<Commit>> {
        if let Some(hash) = self.head_commit()? {
            Ok(Some(self.load_commit(&hash)?))
        } else {
            Ok(None)
        }
    }

    /// List all commits in reverse chronological order.
    pub fn list_commits(&self, limit: usize) -> Result<Vec<Commit>> {
        let mut commits = Vec::new();

        // Start from HEAD and follow parent chain
        if let Some(mut current_hash) = self.head_commit()? {
            while commits.len() < limit {
                let commit = self.load_commit(&current_hash)?;
                let parent = commit.parent_hashes.first().cloned();
                commits.push(commit);

                if let Some(parent_hash) = parent {
                    current_hash = parent_hash;
                } else {
                    break;
                }
            }
        }

        Ok(commits)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use tempfile::tempdir;

    #[test]
    fn test_init_project() {
        let dir = tempdir().unwrap();
        let project = Project::init(dir.path(), "test-project", "testuser").unwrap();

        assert_eq!(project.config.project.name, "test-project");
        assert_eq!(project.config.project.owner, "testuser");
        assert!(dir.path().join("ozzy.toml").exists());
        assert!(dir.path().join(".ozzy/commits").exists());
        assert!(dir.path().join(".ozzy/refs/heads").exists());
        assert!(dir.path().join("data").exists());
        assert!(dir.path().join("transforms").exists());
        let gitignore = fs::read_to_string(dir.path().join(".gitignore")).unwrap();
        assert!(gitignore.contains(".ozzy/"));
        assert!(gitignore.contains("data/*.parquet"));
    }

    #[test]
    fn test_open_project() {
        let dir = tempdir().unwrap();
        Project::init(dir.path(), "test-project", "testuser").unwrap();

        let project = Project::open(dir.path()).unwrap();
        assert_eq!(project.config.project.name, "test-project");
    }

    #[test]
    fn test_find_project() {
        let dir = tempdir().unwrap();
        let subdir = dir.path().join("nested/deep");
        fs::create_dir_all(&subdir).unwrap();

        Project::init(dir.path(), "test-project", "testuser").unwrap();

        let project = Project::find_in(&subdir).unwrap();
        assert_eq!(project.config.project.name, "test-project");
    }

    #[test]
    fn test_not_in_project() {
        let dir = tempdir().unwrap();
        let result = Project::find_in(dir.path());
        assert!(matches!(result, Err(Error::NotInProject)));
    }

    #[test]
    fn test_validate_safe_name_strict_ascii_pattern() {
        assert!(validate_safe_name("main").is_ok());
        assert!(validate_safe_name("branch_1-test").is_ok());
        assert!(validate_safe_name("v1.0.0").is_ok());
        assert!(validate_safe_name(".hidden").is_err());
        assert!(validate_safe_name("has space").is_err());
        assert!(validate_safe_name("../escape").is_err());
    }

    #[test]
    fn test_update_ref_rejects_path_traversal() {
        let dir = tempdir().unwrap();
        let project = Project::init(dir.path(), "test-project", "testuser").unwrap();
        let result = project.update_ref("../../outside", "abc123");
        assert!(matches!(result, Err(Error::InvalidPath(_))));
    }

    #[test]
    fn test_resolve_ref_rejects_path_traversal() {
        let dir = tempdir().unwrap();
        let project = Project::init(dir.path(), "test-project", "testuser").unwrap();
        let result = project.resolve_ref("../../outside");
        assert!(matches!(result, Err(Error::InvalidPath(_))));
    }

    #[test]
    fn test_load_commit_rejects_corrupted_hash() {
        let dir = tempdir().unwrap();
        let project = Project::init(dir.path(), "test-project", "testuser").unwrap();

        // Create a commit via the normal path
        let commit = crate::commit::create_commit(&project, "test commit", "testuser").unwrap();
        project.save_commit(&commit).unwrap();

        // Tamper with the commit file to corrupt it
        let commit_path = project
            .commits_dir()
            .join(format!("{}.json", commit.hash));
        let mut content: serde_json::Value =
            serde_json::from_str(&fs::read_to_string(&commit_path).unwrap()).unwrap();
        content["message"] = serde_json::Value::String("tampered message".to_string());
        fs::write(&commit_path, serde_json::to_string_pretty(&content).unwrap()).unwrap();

        // Loading should fail with hash mismatch
        let result = project.load_commit(&commit.hash);
        assert!(result.is_err(), "load_commit should reject corrupted commit");
        let err_msg = format!("{}", result.unwrap_err());
        assert!(
            err_msg.contains("hash mismatch"),
            "Error should mention hash mismatch, got: {}",
            err_msg
        );
    }
}

File: clients/python/src/ozzydb/project.py
"""
Project loading and management for OzzyDB Python client.
"""

import json
import os
from pathlib import Path
from typing import Any, Optional

import pyarrow.parquet as pq
import toml

from ozzydb.types import (
    DataSourceMeta,
    EndpointMeta,
    PipelineEdge,
    PipelineNode,
    ProjectMeta,
    TransformMeta,
)


class Project:
    """
    Represents a local OzzyDB project.

    Usage:
        project = Project("./my-project")
        print(project.name)
        print(project.endpoints)
    """

    def __init__(self, path: str | Path):
        """
        Load a project from a directory.

        Args:
            path: Path to the project directory (containing ozzy.toml)
        """
        self.root = Path(path).expanduser().resolve()

        if not (self.root / "ozzy.toml").exists():
            raise FileNotFoundError(
                f"Not an OzzyDB project: {self.root} (ozzy.toml not found)"
            )

        self._config = toml.load(self.root / "ozzy.toml")
        self._commit: Optional[dict[str, Any]] = None

    @property
    def name(self) -> str:
        """Project name."""
        return self._config["project"]["name"]

    @property
    def owner(self) -> str:
        """Project owner."""
        return self._config["project"]["owner"]

    @property
    def head(self) -> str:
        """Current HEAD ref."""
        return self._config.get("refs", {}).get("head", "refs/heads/main")

    @property
    def ozzy_dir(self) -> Path:
        """Path to .ozzy directory."""
        return self.root / ".ozzy"

    def _load_latest_commit(self) -> Optional[dict[str, Any]]:
        """Load the latest commit from the refs."""
        if self._commit is not None:
            return self._commit

        # Read the HEAD ref
        head = self.head
        if head.startswith("refs/"):
            ref_path = self.ozzy_dir / head
        else:
            ref_path = self.ozzy_dir / "refs" / "heads" / head

        if not ref_path.exists():
            return None

        commit_hash = ref_path.read_text().strip()
        commit_path = self.ozzy_dir / "commits" / f"{commit_hash}.json"

        if not commit_path.exists():
            return None

        self._commit = json.loads(commit_path.read_text())
        return self._commit

    def _load_staged_data_sources(self) -> dict[str, DataSourceMeta]:
        """Load staged (uncommitted) data sources from data/ directory.

        The CLI scans the data/ directory directly for parquet files,
        so we do the same here for consistency.
        """
        data_dir = self.root / "data"
        if not data_dir.exists():
            return {}

        sources = {}
        for parquet_file in data_dir.glob("*.parquet"):
            name = parquet_file.stem
            # Get file metadata
            stat = parquet_file.stat()
            try:
                # Try to get row count from parquet metadata
                parquet_meta = pq.read_metadata(parquet_file)
                row_count = parquet_meta.num_rows
            except Exception:
                row_count = None

            sources[name] = DataSourceMeta(
                name=name,
                hash=None,  # Hash is computed at commit time
                schema_hash=None,
                path=f"data/{parquet_file.name}",
                row_count=row_count,
                byte_size=stat.st_size,
            )
        return sources

    def _load_staged_transforms(self) -> dict[str, TransformMeta]:
        """Load staged (uncommitted) transforms from transforms/ directory.

        The CLI scans the transforms/ directory for @ozzy.transform decorated
        functions, so we do the same here for consistency.
        """
        transforms_dir = self.root / "transforms"
        if not transforms_dir.exists():
            return {}

        transforms = {}
        for py_file in transforms_dir.glob("*.py"):
            # Parse the file to find @ozzy.transform decorated functions
            try:
                content = py_file.read_text()
                found_transforms = self._parse_transform_file(content, py_file)
                transforms.update(found_transforms)
            except Exception:
                continue

        return transforms

    def _parse_transform_file(
        self, content: str, file_path: Path
    ) -> dict[str, TransformMeta]:
        """Parse a Python file for @ozzy.transform decorated functions."""
        import re

        transforms = {}
        lines = content.split("\n")

        in_decorator = False
        for i, line in enumerate(lines):
            stripped = line.strip()
            if stripped.startswith("@ozzy.transform"):
                in_decorator = True
            elif in_decorator and stripped.startswith("def "):
                # Extract function name
                match = re.match(r"def\s+(\w+)\s*\(", stripped)
                if match:
                    func_name = match.group(1)
                    rel_path = f"transforms/{file_path.name}"
                    transforms[func_name] = TransformMeta(
                        name=func_name,
                        hash=None,  # Hash is computed at commit time
                        runtime="python",
                        source_path=rel_path,
                        function_name=func_name,
                        lockfile_hash=None,
                        params_schema={},
                        reproducible=True,
                        input_schema=None,
                        output_schema=None,
                    )
                in_decorator = False

        return transforms

    def _load_staged_endpoints(self) -> dict[str, dict[str, Any]]:
        """Load staged (uncommitted) endpoints."""
        staged_dir = self.ozzy_dir / "staged_endpoints"
        if not staged_dir.exists():
            return {}

        endpoints = {}
        for f in staged_dir.glob("*.json"):
            data = json.loads(f.read_text())
            endpoints[f.stem] = data
        return endpoints

    def _load_staged_endpoint_deletions(self) -> set[str]:
        """Load endpoint names staged for deletion."""
        staged_dir = self.ozzy_dir / "staged_endpoints"
        if not staged_dir.exists():
            return set()

        deletions = set()
        for marker in staged_dir.glob("*.deleted"):
            deletions.add(marker.stem)
        return deletions

    @property
    def data_sources(self) -> dict[str, DataSourceMeta]:
        """All data sources (committed + staged)."""
        sources = {}

        # Load from latest commit
        commit = self._load_latest_commit()
        if commit:
            for name, data in commit.get("data_sources", {}).items():
                sources[name] = DataSourceMeta(
                    name=data["name"],
                    hash=data["hash"],
                    schema_hash=data["schema_hash"],
                    path=data["path"],
                    row_count=data.get("row_count"),
                    byte_size=data.get("byte_size"),
                )

        # Override with staged
        sources.update(self._load_staged_data_sources())
        return sources

    @property
    def transforms(self) -> dict[str, TransformMeta]:
        """All transforms (committed + staged)."""
        transforms = {}

        # Load from latest commit
        commit = self._load_latest_commit()
        if commit:
            for name, data in commit.get("transforms", {}).items():
                transforms[name] = TransformMeta(
                    name=data["name"],
                    hash=data["hash"],
                    runtime=data["runtime"],
                    source_path=data["source_path"],
                    function_name=data["function_name"],
                    lockfile_hash=data["lockfile_hash"],
                    params_schema=data.get("params_schema", {}),
                    reproducible=data.get("reproducible", True),
                    input_schema=data.get("input_schema"),
                    output_schema=data.get("output_schema"),
                )

        # Override with staged
        transforms.update(self._load_staged_transforms())
        return transforms

    @property
    def endpoints(self) -> list[str]:
        """List of endpoint names."""
        endpoint_names = set()
        staged = self._load_staged_endpoints()
        staged_deletions = self._load_staged_endpoint_deletions()

        # From commit
        commit = self._load_latest_commit()
        if commit:
            endpoint_names.update(commit.get("endpoints", {}).keys())

        # From staged
        endpoint_names.update(staged.keys())

        # Staged deletions hide committed endpoints, unless the endpoint is
        # simultaneously restaged with a JSON definition.
        endpoint_names.difference_update(staged_deletions - set(staged.keys()))

        return sorted(endpoint_names)

    def get_endpoint(self, name: str) -> EndpointMeta:
        """
        Get metadata for a specific endpoint.

        Args:
            name: Endpoint name

        Returns:
            EndpointMeta with full DAG information
        """
        # Check staged first
        staged = self._load_staged_endpoints()
        staged_deletions = self._load_staged_endpoint_deletions()
        if name in staged:
            endpoint_data = staged[name]
        elif name in staged_deletions:
            raise KeyError(f"Endpoint '{name}' is staged for deletion")
        else:
            # Check committed
            commit = self._load_latest_commit()
            if not commit or name not in commit.get("endpoints", {}):
                raise KeyError(f"Endpoint '{name}' not found")
            endpoint_data = commit["endpoints"][name]

        # Parse nodes
        nodes = [
            PipelineNode(
                node_name=n["node_name"],
                transform_name=n["transform_name"],
                params=n.get("params", {}),
            )
            for n in endpoint_data.get("nodes", [])
        ]

        # Parse edges
        edges = [
            PipelineEdge(
                target_node=e["target_node"],
                input_name=e["input_name"],
                source_type=e["source_type"],
                source_ref=e["source_ref"],
                external_owner=e.get("external_owner"),
                external_project=e.get("external_project"),
                external_endpoint=e.get("external_endpoint"),
                external_commit_hash=e.get("external_commit_hash"),
            )
            for e in endpoint_data.get("edges", [])
        ]

        return EndpointMeta(
            name=name,
            nodes=nodes,
            edges=edges,
            data_sources=self.data_sources,
            transforms=self.transforms,
            hash=endpoint_data.get("dag_hash"),
        )

    def get_data_path(self, source_name: str) -> Path:
        """Get the full path to a data source file."""
        source = self.data_sources.get(source_name)
        if not source:
            raise KeyError(f"Data source '{source_name}' not found")
        return self.root / source.path

    def get_schema(self, source_name: str):
        """Get the Arrow schema of a data source."""
        path = self.get_data_path(source_name)
        return pq.read_schema(path)

    def meta(self) -> ProjectMeta:
        """Get full project metadata."""
        return ProjectMeta(
            name=self.name,
            owner=self.owner,
            root=str(self.root),
            head=self.head,
            data_sources=self.data_sources,
            transforms=self.transforms,
            endpoints=self.endpoints,
        )
